packages/subgraph-service/contracts/libraries/IndexingAgreement.sol

```diff
@@ -0,0 +1,459 @@
+        require(_isActive(wrapper), IndexingAgreementNotActive(agreementId));
+        require(
+            _directory().recurringCollector().isAuthorized(wrapper.collectorAgreement.payer, msg.sender),
+            IndexingAgreementNonCancelableBy(wrapper.collectorAgreement.payer, msg.sender)
+        );
+        _cancel(
+            self,
+            agreementId,
+            wrapper.agreement,
+            wrapper.collectorAgreement,
+            IRecurringCollector.CancelAgreementBy.Payer
+        );
+    }
+
+    function collect(
+        Manager storage self,
+        mapping(address allocationId => Allocation.State allocation) storage allocations,
+        bytes16 agreementId,
+        bytes memory data
+    ) external returns (address, uint256) {
+        AgreementWrapper memory wrapper = _get(self, agreementId);
+        Allocation.State memory allocation = allocations.requireValidAllocation(
+            wrapper.agreement.allocationId,
+            wrapper.collectorAgreement.serviceProvider
+        );
+        require(_isActive(wrapper), IndexingAgreementNotActive(agreementId));
+
+        require(
+            wrapper.agreement.version == IndexingAgreementVersion.V1,
+            InvalidIndexingAgreementVersion(wrapper.agreement.version)
+        );
+
+        (uint256 entities, bytes32 poi, uint256 poiEpoch) = Decoder.decodeCollectIndexingFeeDataV1(data);
+
+        uint256 expectedTokens = (entities == 0 && poi == bytes32(0))
+            ? 0
+            : _tokensToCollect(self, agreementId, wrapper.collectorAgreement, entities);
+
+        uint256 tokensCollected = _directory().recurringCollector().collect(
+            IGraphPayments.PaymentTypes.IndexingFee,
+            abi.encode(
+                IRecurringCollector.CollectParams({
+                    agreementId: agreementId,
+                    collectionId: bytes32(uint256(uint160(wrapper.agreement.allocationId))),
+                    tokens: expectedTokens,
+                    dataServiceCut: 0
+                })
+            )
+        );
+
+        emit IndexingFeesCollectedV1(
+            wrapper.collectorAgreement.serviceProvider,
+            wrapper.collectorAgreement.payer,
+            agreementId,
+            wrapper.agreement.allocationId,
+            allocation.subgraphDeploymentId,
+            _graphDirectory().graphEpochManager().currentEpoch(),
+            tokensCollected,
+            entities,
+            poi,
+            poiEpoch
+        );
+
+        return (wrapper.collectorAgreement.serviceProvider, tokensCollected);
+    }
+
+    function get(Manager storage self, bytes16 agreementId) external view returns (AgreementWrapper memory) {
+        AgreementWrapper memory wrapper = _get(self, agreementId);
+        require(wrapper.collectorAgreement.dataService == address(this), IndexingAgreementNotActive(agreementId));
+
+        return wrapper;
+    }
+
+    function _getManager() internal pure returns (Manager storage manager) {
```

Comment:
This is a good idea but I recommend using ERC7201 like the OZ contracts: https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/utils/PausableUpgradeable.sol#L19-L31